src/ORM/Association/HasMany.php

```diff
@@ -134,6 +181,63 @@ public function saveAssociated(EntityInterface $entity, array $options = [])
     }
 
     /**
+     * Deletes/sets null the related objects according to the dependency between source and targets and foreign key nullability
+     * Skips deleting records present in $remainingEntities
+     *
+     * @param array $properties array of foreignKey properties
+     * @param EntityInterface $entity the entity which should have its associated entities unassigned
+     * @param Table $target The associated table
+     * @param array $remainingEntities Entities that should not be deleted
+     * @return void
+     */
+    protected function _unlinkAssociated(array $properties, EntityInterface $entity, Table $target, array $remainingEntities)
+    {
+        $primaryKey = (array)$target->primaryKey();
+        $mustBeDependent = (!$this->_foreignKeyAcceptsNull($target, $properties) || $this->dependent());
+        $conditions = [
+            'AND' => [
+                'NOT' => [
+                    'AND' => array_map(
```

Comment:
You have to see it as a negative statement.
It will delete associated data that does not contain any of the ids provided in the list:

not (id=1,id=2,id=3) = not id=1 AND not id=2 AND not id=3. It means that if one of the conditions is truth (it means, the associated entity has one of the ids) it will not be deleted.